Updates from new fork

ericlagergren opened the pull request:
Hi! We recently forked your library to and revamped it to fit our needs better. Just to let you know about it, I'm creating this PR. Hope you find some of the changes useful!

ericlagergren committed:
misc: remove most allocs from encoding/{Encoder,Decoder}, simplify code

ericlagergren committed:
recordings: fix broken chunked recording

ericlagergren committed:
recordings: update recording files

Also: remove alloc from chunkWriter.WriteString

ericlagergren committed:
separate recorder from driver and make two separate drivers

ericlagergren committed:
misc: move to a more driver compliant structure

ericlagergren committed:
misc: migrate to new pool type

ericlagergren committed:
misc: migrate to new pooling

ericlagergren committed:
misc: many updates, mostly moving to copy of database.DB interface

ericlagergren committed:
misc: fix stmt issues, tx pending

ericlagergren committed:
misc: finalize TX

ericlagergren committed:
update LICENSE and README back to pre-fork

johnnadratowski commented:
@ericlagergren - Wow, you guys really ripped this apart! 👍  Happy to see people customizing this library to their needs.

Thank you for submitting this PR.  It is a lot to go through!

Do you have any specific features you added that you think I should consider from this PR? Any fixes?

ericlagergren commented:
you're welcome! I appreciate you writing the library in the first place—it I'm glad I did not have to write something from scratch.

There are several big changes, but the two that'd provide the biggest bang for your buck are the 'encoding' package and the conn type.

Encoding:

Decoding messages is inefficient—it reads everything has unnecessary allocations, and so on. I got rid of pretty much any unnecessary allocation I could find, only reading one "type" at a time. Encoding is similar, and I removed some []byte(string) conversions.

Conn type:

I was iffy on having a recorder baked into the conn type, so I split it into two different types. Also, `boltConn` was gigantic so I removed superfluous fields. Instead of creating multiple conns, I mimicked sql.DB and made Query, Exec, etc. all methods. Those modifications allowed me to move the stmt and tx types out from the conn type.

The modifying conn was biggest change and probably not 100% clear off the bat.

ericlagergren commented:
oh! one other thing. I noticed in a couple places you check to see if *something* (a string or slice) is too large and if it is you print it out.

I'd recommend removing that—printing a 3GB+ string is probably not a good idea.
